services/configuration/src/config_api.rs

```diff
@@ -70,54 +72,52 @@ pub struct PublicConfigApi {
-            hash: actual_cfg.hash(),
-            config: actual_cfg,
-        };
-        Ok(res)
+    fn get_config_with_proofs(&self, config: StoredConfiguration) -> ApiResponseConfigHashInfo {
+        let propose = ConfigurationSchema::new(&self.blockchain.snapshot())
+            .get_propose(&config.hash())
+            .map(|p| p.hash());
+        let votes = self.get_votes_for_propose(&config.hash());
+        ApiResponseConfigHashInfo {
+            hash: config.hash(),
+            config,
+            propose,
+            votes,
+        }
     }
 
-    fn get_following_config(&self) -> Result<Option<ApiResponseConfigHashInfo>, ApiError> {
+    fn get_actual_config(&self) -> ApiResponseConfigHashInfo {
+        let snapshot = self.blockchain.snapshot();
+        let configuration_schema = Schema::new(&snapshot);
+        let actual_cfg = configuration_schema.actual_configuration();
+        self.get_config_with_proofs(actual_cfg)
+    }
 
-        let following_cfg = Schema::new(&self.blockchain.snapshot()).following_configuration();
-        let res = following_cfg.map(|cfg| {
-            ApiResponseConfigHashInfo {
-                hash: cfg.hash(),
-                config: cfg,
-            }
-        });
-        Ok(res)
+    fn get_following_config(&self) -> Option<ApiResponseConfigHashInfo> {
+        Schema::new(&self.blockchain.snapshot())
+            .following_configuration()
+            .map(|following_cfg| self.get_config_with_proofs(following_cfg))
     }
 
-    fn get_config_by_hash(&self, hash: &Hash) -> Result<ApiResponseConfigInfo, ApiError> {
+    fn get_config_by_hash(&self, hash: &Hash) -> ApiResponseConfigInfo {
         let snapshot = self.blockchain.snapshot();
         let general_schema = Schema::new(&snapshot);
         let committed_config = general_schema.configs().get(hash);
 
         let configuration_schema = ConfigurationSchema::new(&snapshot);
         let propose = configuration_schema.propose_data_by_config_hash().get(hash);
-        let res = ApiResponseConfigInfo {
-            committed_config: committed_config,
-            propose: propose,
-        };
-        Ok(res)
+        ApiResponseConfigInfo {
+            committed_config,
+            propose,
+        }
     }
 
-
-    fn get_votes_for_propose(&self, cfg_hash: &Hash) -> Result<ApiResponseVotesInfo, ApiError> {
+    fn get_votes_for_propose(&self, config_hash: &Hash) -> ApiResponseVotesInfo {
         let snapshot = self.blockchain.snapshot();
         let configuration_schema = ConfigurationSchema::new(&snapshot);
-        let res = match configuration_schema.propose_data_by_config_hash().get(
-            cfg_hash,
-        ) {
-            None => None,
-            Some(_) => Some(configuration_schema.get_votes(cfg_hash)),
-        };
-
-        Ok(res)
+        configuration_schema
+            .propose_data_by_config_hash()
+            .get(config_hash)
+            .map(|_| configuration_schema.get_votes(config_hash))
```

Comment:
Seems highly suspicious that it just ignores `propose_data`. I think it is required for some reason, but perhaps a code comment, explaining why we need this lookup, would make this slightly less mysterious? :)

Reply:
I would love to see such comment too... :smile:

@alekseysidorov @slowli @vldm 
Can anyone remember why it is needed? Is it just a refactoring artifact that can be safely removed?

Reply:
I think I understand the general idea: the code checks if there exists a proposal with the given `config_hash` and returns votes for it only if it exists. The reasoning for a preliminary existence check is that `get_votes()` method uses a group of indexes distinguished by a prefix to store / retrieve data, and I think (not 100% sure) that there are no methods in Exonum API to check whether a particular index within the family exists.

Still, if I'm correct, the code could be refactored to use `contains(config_hash)` instead of `get`.